src/board_builder/board_builder.py

```diff
@@ -0,0 +1,258 @@
+        """ Given a matrix transformation, find the location of the four corners """
+        corners_reference = np.zeros((4, 4))
+        for i in range(4):
+            corners_reference[i] = T_matrix @ local_corners[i]
+
+        corners_reference = corners_reference[:, :3]
+        return corners_reference
+
+    def _calculate_relative_transform(self, T1, T2):
+        """ Given transform T1 from reference to marker 1, and transfrom T2 from reference to marker 2, calculate the
+        transform from T1 to T2"""
+        T1_inv = np.linalg.inv(T1)
+        T_relative = np.dot(T1_inv, T2)
+        return T_relative
+
+    def _estimate_reference_to_occluded(self, T1, T2):
+        """ Given transform T1 from reference to visible marker, and transform T2 from visible marker to occluded
+        marker, calculate the transform from the reference to the occluded marker"""
+        T_reference_to_occluded = np.dot(T1, T2)
+        return T_reference_to_occluded
+
+    def _expand_matrix(self):
+        """ Adds one row and one column to the matrix and initializes them to None """
+        size = len(self._relative_pose_matrix) + 1
+        new_matrix = [[None for _ in range(size)] for _ in range(size)]
+        for i in range(size - 1):
+            for j in range(size - 1):
+                new_matrix[i][j] = self._relative_pose_matrix[i][j]
+        self._relative_pose_matrix = new_matrix
+
+    def _filter_data(self, data):
+        """ Filters the detector data to only keep the markers snapshots that appear in more than one detector """
+        label_counts = defaultdict(int)
+        for snapshots in data.values():
+            for snapshot in snapshots:
+                label_counts[snapshot.label] += 1
+
+        filtered_data = {}
+        for key, snapshots in data.items():
+            filtered_snapshots = [
+                snapshot for snapshot in snapshots
+                if label_counts[snapshot.label] >= 2
+            ]
+            if filtered_snapshots:
+                filtered_data[key] = filtered_snapshots
+        return filtered_data
+
+    def _find_matrix_input_index(self, pose_uuid, other_pose_uuid):
+        """ Given two pose uuids, return their index location in the relative pose matrix """
+
+        pose_index = -1
+        other_pose_index = -1
+
+        for index in self._index_to_marker_uuid:
+            if self._index_to_marker_uuid[index] == pose_uuid:
+                pose_index = index
+            if self._index_to_marker_uuid[index] == other_pose_uuid:
+                other_pose_index = index
+
+        if pose_index == -1:
+            raise ValueError(f"Pose UUID {pose_uuid} not found in the index mapping.")
+        if other_pose_index == -1:
+            raise ValueError(f"Pose UUID {other_pose_uuid} not found in the index mapping.")
+
+        return pose_index, other_pose_index
+
+    def _solve_pose(self, detector_data, timestamp):
```

Comment:
Timestamp is a string: str(datetime.datetime.now()) which should follow ISO. The reason  am using string is that I think some part of pose solver uses string datatime instead of the actual datetime.